### ISSUE TITLE
PD doesn't work with the "new" jack2 on macOS

### ISSUE THREAD

**Issue by dyfer** [pure-data/pure-data#1190]
PureData is not able to use the "new" jack2 (as of Q4 2020) on macOS, AFAICT. Tested with a [recent jack2](https://github.com/jackaudio/jack2-releases/releases) on macOS 10.15 (Catalina). After selecting `jack` as the backend, PD reports that jack is not installed.

IIUC, the issue is that PD is linked with the jack "framework" that was provided with JackOSX, but is not provided with the current jack2.

I'm mostly logging this to raise awareness about the issue with jack(2) development on macOS and backward compatibility. I also reported compatibility issue in the jack2 repo: https://github.com/jackaudio/jack2/issues/640

**Comment by danomatika** [comment]
>PureData is not able to use the "new" jack2 (as of Q4 2020) on macOS, AFAICT.

I assume you are referring to the precompiled versions you can download, which is true as the older framework is still weak linked by default.

 if you build Pd yourself, you can disable this and explicitly link to jack installed on the system, ie. via homebrew:

    ./configure   --disable-jack-framework --enable-jack

The problem with this, is that it's not distributable as the same version of jack has to be present on the system running Pd. Currently, Pd is built for such a wide range of systems, going back to 10.6, that this is a bit more difficult than what the older JackOSX lib provided.

The basic info on this is in the [INSTALL.txt](https://github.com/pure-data/pure-data/blob/master/INSTALL.txt#L233), however most general users don't need to look into it.

**Comment by danomatika** [comment]
I did some checking and can confirm building Pd without the framework allows me to connect to the Homebrew jack server.

The issue is more that we need some sort of replacement for the framework that allows us to distribute Pd with jack enabled on macOS *without* requiring it on user systems. Maybe this is actually easy, [ala `-weak_library jack`](https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/BPFrameworks/Concepts/WeakLinking.html#//apple_ref/doc/uid/20002378-107026) but I wasn't able to get libtool to link with it set.

**Comment by dyfer** [comment]
I have limited experience with this, but I came across information implying that weak linking to a framework is easier than to a library. Again, my experience is very limited. AFAIU current jack2 provides only the shared library, and not the framework.
In the linked thread @falkTX mentioned https://github.com/x42/weakjack/ as a possible library to easily weak link jack, I have not tried it though.

**Comment by danomatika** [comment]
We are using jack with another project at work and I did some testing. Our project and Pd compiled with `-ljack` and the current official pre-built Jack install work fine together. Also, running both apps on another mac without jack installed also work fine and simply complain about the jack library not being found, no crashes. Apparently, we don't need to use `-weak_library` and the compiler wasn't letting me using it anyway.

I think the path forward for this with Pd is to change the default to try to link jack normally without the framework. Then Miller can explicitly link to the framework for the 3 bit builds intended for older systems while the 64 builds for newer systems can use jack2.

**Comment by danomatika** [comment]
Ping @reduzent @millerpuckette @umlaeute

**Comment by danomatika** [comment]
Here is a Pd 0.51-3 build linked to jack2: http://docs.danomatika.com/pdbuilds/0.52/Pd-0.51.3-jack.zip

It should simply show the error if you try to switch to jack on a system without jack installed or with the old JackOSX build. To use it, download the [official jack installer for macOS](https://jackaudio.org/downloads/). I believe it should also work with jack2 installed via homebrew as it will also be placed in /usr/local and the jack API is pretty stable. In any case, it's good to be abel to simply point users to the official jack website and their installer.

**Comment by jakeatwork** [comment]
@danomatika thanks for the pointer - and I appreciate offering the build, but decided to build it myself and it worked fine and was pretty easy. Here's the steps I took:

1. `brew install jack` from terminal
2.  `brew install qjackctl`
3. make sure you start jack with `brew services start jack`
4. on https://puredata.info/downloads/pure-data, click  "Get Pure Data for All platforms (Source code)" and save it to a good location (that link might fail in some browsers because it's `http` only, so copy/paste the url into a new browser tab and it should download fine)
5. unzip the file you just downloaded
6. back in terminal, `cd` into the unzipped directory
7. `./autogen.sh`
8 `./configure --disable-jack-framework --enable-jack` (as @danomatika pointed out above. you can read more at: https://github.com/pure-data/pure-data/blob/master/INSTALL.txt#L233)
9. if that's all you need to configure, then run `make`
10. `make app`
11. `mv Pd-0.51.3.app /Applications/` (assuming that's the Pd version you have)

i think that's it, but it's all working great now on catalina. appreciate the pointers!

**Comment by reduzent** [comment]
> We are using jack with another project at work and I did some testing. Our project and Pd compiled with `-ljack` and the current official pre-built jack2 install work fine together. Also, running both apps on another mac without jack installed also works fine and they simply complain about the jack library not being found, no crashes. Apparently, we don't need to use `-weak_library` and the compiler wasn't letting me using it anyway.
> 
> I think the path forward for this with Pd is to change the default to try to link jack normally without the framework. Then Miller can explicitly link to the framework for the 32 bit builds intended for older systems while the 64 builds for newer systems can use jack2.

Your experiences are congruent with mine. When no JACK is installed or running, I only get:

```
JACK: couldn't connect to server, is JACK running?
```
Pd keeps running fine.

**Comment by dyfer** [comment]
Another option would be to use [weakjack](https://github.com/x42/weakjack). JackTrip started using it recently and it works well with macOS and Windows with a range of Jack versions.

**Comment by umlaeute** [comment]
weakjack looks interesting, but alas it's GPL-licensed (and Pd is BSD3)

**Comment by danomatika** [comment]
I don’t see why we need the separate weakjacknlib when, in my testing, building with the official jack lib build itself should work fine on other systems. The caveat, as I’ve mentioned previously, is that the build system has to be sufficiently modern to use the new lib and builds will most likely not work on much older systems. I don’t see this as a major problem considering said systems can’t install the jack2 lib builds anyway. My opinion is to keep weakling support for old systems which can support jackOSX and build with jack2 for new systems.

enohp ym morf tnes
-----------
Dan Wilcox
danomatika.com
robotcowboy.com

**Comment by umlaeute** [comment]
actually i am appealed by weakling weak-linking for *all* platforms.

**Comment by danomatika** [comment]
Hah "weakling" was a bad auto-correction.

> On Sep 29, 2021, at 4:30 PM, umläute ***@***.***> wrote:
> 
> weakling

--------
Dan Wilcox
@danomatika <http://twitter.com/danomatika>
danomatika.com <http://danomatika.com/>
robotcowboy.com <http://robotcowboy.com/>

**Comment by dyfer** [comment]
> I don’t see why we need the separate weakjack lib when, in my testing, building with the official jack lib build itself should work fine on other systems. 

The point of using weak linking of jack (regardless if/which which library is being used for that) is to not require users to install jack in order to just _run_ PD, since PD can indeed operate without jack (using portaudio IIUC).

**Comment by danomatika** [comment]
Again, in my testing, building with the jack2 distribution gives you a binary that you can run on a system without jack installed. I can test this again to make sure but it’s the same approach I use for a macOS application we have available to download from our work website.

enohp ym morf tnes
-----------
Dan Wilcox
danomatika.com
robotcowboy.com

**Comment by dyfer** [comment]
> Again, in my testing, building with the jack2 distribution gives you a binary that you can run on a system without jack installed. I can test this again to make sure but it’s the same approach I use for a macOS application we have available to download from our work website.

Ah, I see. This was not my experience working with jack on macos with JackTrip, but if that works here that would be great.

**Comment by danomatika** [comment]
I just tried my build from November (from [this post](https://github.com/pure-data/pure-data/issues/1190#issuecomment-729601300)) on my wife's 2015 MBP 13 running macOS 11 *without* jack installed and Pd starts and runs fine. If you select jack there is an error print saying that jack does not appear to be installed. Using portaudio works fine.

**Comment by dyfer** [comment]
@danomatika FYI when I compiled JackTrip against jack2 1.9.16 (but not later versions), I was able to run it with both newer versions of jack (1.9.16 and later) as well as the old JackOSX. JackOSX does provide the libraries in `/usr/local/lib` - it provides `libjack.dylib` and `libjack.0.dylib`, though it does not provide `libjack.0.1.0.dylib`. IIUC jack2 1.9.17 and later changed the library versioning so that an application linked against 1.9.17 failed to open with JackOSX, but it _would run_ with all versions of jack if it was linked against 1.9.16. 

I'm sharing this in case it would be indeed possible to create a single PD build that would be compatible with all jack versions.

**Comment by umlaeute** [comment]
(note: i'm translating that from the linux world; i haven't actually tested that it is the same on macOS, but am pretty confident)

> it does not provide libjack.0.1.0.dylib

that shouldn't be a problem, as typically `libjack.0.dylib` would be a symlink to `libjack.0.1.0.dylib` (actually one of `libjack.0.1.*.dylib`).

according to [semantic versioning](https://semver.org/), all `libjack.0.dylib` variants should be backwards compatible.
however, `libjack.1.dylib` would introduce breaking changes.
because of the breaking changes, applications that dynamically link against `libjack.0.dylib` cannot be expected to simply work with `libjack.1.dylib` (in some cases they do, but that is mostly luck) - and as such i don't think you can have a version of Pd that just "works with all versions of JACK".

i guess, some macOS construct like "Frameworks" (as opposed to ordinary "dynamic libraries") might help here, but afaics JACK development ditched the use of Frameworks.

> IIUC jack2 1.9.17 and later changed the library versioning [... ]

of course there's nothing preventing humans from forcefully breaking things.

**Comment by umlaeute** [comment]
also note that in Debian, policy says that binaries should always be linked against *JACK1*, as this will make the application work with both JACK1 **and** JACK2, whereas this might not be true when doing it the other way round.

the policy is probably more than a decade old and things might have changed in the meantime; otoh it still seems to work, otherwise there would be more bug-reports.
also this is of course the *Debian* policy and might not apply to *macOS* (for one thing, in Debian we have library maintainers that actually care that the entire ecosystems continues to work throughout different library versions).

**Comment by dyfer** [comment]
The reason I've pointed out versions provided on macOS is that versioning (as in version indicated in the library) has changed between 1.9.16 and 1.9.17. 1.9.17 also started providing `libjack.0.1.0.dylib` IIRC. The API has not changed though.

I don't think the note on `libjack.1.dylib` is relevant - I'm not aware this version is used anywhere (unless my information is outdated).

As the API has not changed, I think it should be possible to provide a build that works with all currently available Jack versions on macOS - unless I'm missing something.

As for jack1 - it was feasible to link to it at some point earlier on macOS, as for a while jack1 was provided with homebrew. I have compiled JackTrip (which uses Jack) against it and IIRC that also made it work with both JackOSX and newer jack. But again, I'm pretty sure I was able to achieve such compatibility linking against Jack 1.9.16 as well.

It just occurred to me - I'm not sure this was pointed out earlier in the discussion: the "old" jack on macOS, i.e. JackOSX is indeed jack2, not jack1.
